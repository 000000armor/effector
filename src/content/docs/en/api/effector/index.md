---
id: effector
title: API Reference
---

# List of the methods

## Unit creators

- [createEvent](./createEvent)
- [createStore](./createStore)
- [createEffect](./createEffect)
- [createDomain](./createDomain)

## Unit definitions

- [Event](./Event)
- [Effect](./Effect)
- [Store](./Store)
- [Domain](./Domain)
- [Scope](./Scope)

## Common methods

- [combine(...stores, f)](./combine)
- [attach({effect, mapParams, source?})](./attach)
- [createApi(store, api)](./createApi)
- [forward({from, to})](./forward)
- [merge([eventA, eventB])](./merge)
- [sample({clock, source, fn, target})](./sample)
- [guard({source, filter, target})](./guard)
- [split(event, cases)](./split)
- [fromObservable(observable)](./fromObservable)

## Fork API

- [fork](./fork)
- [serialize](./serialize)
- [hydrate](./hydrate)
- [allSettled](./allSettled)
- [scopeBind](./scopeBind)

## Plugins

- [effector/babel-plugin](./babel-plugin)
- [@effector/swc-plugin](https://github.com/effector/swc-plugin)

## Utilities

- [is](./is)

## Low level api

- [clearNode](./clearNode)
- [withRegion](./withRegion)
- [launch](./launch)
