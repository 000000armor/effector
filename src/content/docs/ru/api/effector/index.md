---
title: Справочник по API
description: Перечень основных методов API, по группам
lang: ru
---

Перечень основных методов API, по группам

## Создание юнитов

- [createEvent](./createEvent)
- [createStore](./createStore)
- [createEffect](./createEffect)
- [createDomain](./createDomain)

## Типы юнитов

- [_Event_](./Event)
- [_Effect_](./Effect)
- [_Store_](./Store)
- [_Domain_](./Domain)
- [_Scope_](./Scope)

## Основные методы библиотеки

- [combine(...stores, f)](./combine)
- [attach({effect, mapParams, source?})](./attach)
- [createApi(store, api)](./createApi)
- [forward({from, to})](./forward)
- [merge([eventA, eventB])](./merge)
- [sample({clock, source, fn, target})](./sample)
- [guard({source, filter, target})](./guard)
- [split(event, cases)](./split)
- [fromObservable(observable)](./fromObservable)

## Fork API

- [fork](./fork)
- [serialize](./serialize)
- [hydrate](./hydrate)
- [allSettled](./allSettled)
- [scopeBind](./scopeBind)

## Плагины для компилятора

- [effector/babel-plugin](./babel-plugin)
- [@effector/swc-plugin](https://github.com/effector/swc-plugin)

## Служебные функции

- [is](./is)

## Низкоуровневый API

- [clearNode](./clearNode)
- [withRegion](./withRegion)
- [launch](./launch)
