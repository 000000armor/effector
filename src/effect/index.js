//@flow

export type {Effect} from './index.h'

export {effectFabric} from './effectFabric'
export {createEffect} from './createEffect'
