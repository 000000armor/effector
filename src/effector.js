//@flow

import type {Store} from 'redux'
import {Stream, from} from 'most'
import {async as subject, type Subject} from 'most-subject'

import type {ID, Tag, Domain, Effect, Event} from './index.h'
import {counter, toTag} from './index.h'

export function createDomain<State>(
  store: Store<State>,
  domainName: string = '~'
): Domain<State> {
  const redux: any = store
  const state$: Stream<State> = from(redux)
    .map(() => store.getState())
    .multicast()
  return DomainConstructor(
    domainName,
    redux.dispatch,
    store.getState,
    state$,
  )
}

export function domain<State>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
): Domain<State> {
  return DomainConstructor(
    domainName,
    dispatch,
    getState,
    state$,
  )
}


const nextPayloadID = counter()
const nextEventID = counter()

function DomainConstructor<State>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events = new Map
): Domain<State> {
  return {
    effect<Params, Done, Fail>(
      name: string
    ): Effect<Params, Done, Fail, State> {
      return EffectConstructor(
        domainName,
        dispatch,
        getState,
        state$,
        events,
        name
      )
    },
    domain(name: string) {
      return DomainConstructor(
        [domainName, name].join('/'),
        dispatch,
        getState,
        state$,
        events,
      )
    },
    event<Payload>(
      name: string
    ): Event<Payload, State> {
      return EventConstructor(
        domainName,
        dispatch,
        getState,
        state$,
        events,
        name
      )
    }
  }
}

function EffectConstructor<State, Params, Done, Fail>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events,
  name: string
): Effect<Params, Done, Fail, State> {
  const handlers = new Set
  const eventID = nextEventID()
  const nextSeq = counter()
  const action$: Subject<Params> = subject()
  const getType = () => toTag(domainName, name)
  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  let using: (params: Params) => Promise<Done> = never
  function use(thunk: (params: Params) => Promise<Done>) {
    using = thunk
  }
  const create = (payload: Params) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const effect = (params: Params) => {
    let run = () => {
      run = () => {}
      using(params).then(
        result => {
          const data = {params, result}
          dispatch(done(data))
          resolve(data)
        },
        error => {
          const data = {params, error}
          dispatch(fail(data))
          reject(data)
        },
      )
    }
    const runner = (dispatchHook) => {
      if (canDispatch) {
        const dispatcher = dispatchHook
          ? dispatchHook
          : dispatch
        canDispatch = false
        dispatcher(result)
        for (const handler of handlers) {
          handler(params, getState())
        }
        action$.next(params)
        run()
      }
    }
    let resolve = (val) => {}
    let reject = (val) => {}
    const doneP: Promise<{params: Params, result: Done}> = new Promise(rs => resolve = rs)
    const failP: Promise<{params: Params, error: Fail}> = new Promise(rs => reject = rs)
    const toSend = Promise.resolve(params)
    const result = create(params)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        runner(dispatchHook)
        return toSend
      },
      done() {
        runner()
        return doneP
      },
      fail() {
        runner()
        return failP
      },
      promise() {
        runner()
        return Promise.race([
          doneP,
          failP.then(data => Promise.reject(data)),
        ])
      }
    }
  }
  const done: Event<{params: Params, result: Done}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'done'].join(' ')
  )
  const fail: Event<{params: Params, error: Fail}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'fail'].join(' ')
  )
  function watch<R>(fn: (params: Params, state: State) => R) {
    handlers.add((params, state) => {
      const result = fn(params, state)
      safeDispatch(result, dispatch)
    })
  }
  function epic<R>(
    handler: (
      data$: Stream<Params>,
      state$: Stream<State>
    ) => Stream<R>
  ): Stream<R> {
    const result = handler(action$, state$)
      .multicast()
    result.observe(data => safeDispatch(data, dispatch))
    return result
  }
  effect.use = use
  effect.getType = getType
  effect.done = done
  effect.fail = fail
  effect.watch = watch
  effect.epic = epic
  return effect
}

function safeDispatch(data, dispatch) {
  if (
    typeof data === 'object'
    && data != null
    && typeof data.type === 'string'
  ) {
    dispatch(data)
  }
}

function EventConstructor<State, Payload>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events,
  name: string
): Event<Payload, State> {
  const eventID = nextEventID()
  const nextSeq = counter()
  const getType = () => toTag(domainName, name)
  const handlers = new Set
  function watch<R>(
    fn: (params: Payload, state: State) => R
  ) {
    handlers.add(fn)
  }
  const action$: Subject<Payload> = subject()
  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  function epic<R>(
    handler: (
      data$: Stream<Payload>,
      state$: Stream<State>
    ) => Stream<R>
  ): Stream<R> {
    const result = handler(action$, state$)
      .multicast()
    result.observe(data => safeDispatch(data, dispatch))
    return result
  }
  const create = (payload: Payload) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const eventInstance = (payload: Payload) => {
    const toSend = Promise.resolve(payload)
    const result = create(payload)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        if (canDispatch) {
          const dispatcher = dispatchHook
            ? dispatchHook
            : dispatch
          canDispatch = false
          dispatcher(result)
          for (const handler of handlers) {
            handler(payload, getState())
          }
          action$.next(payload)
        }
        return toSend
      },
    }
  }
  eventInstance.getType = getType
  //$off
  eventInstance.toString = getType
  eventInstance.watch = watch
  eventInstance.epic = epic

  events.set(getType(), eventInstance)
  return eventInstance
}


function never(props: any): Promise<any> {
  console.warn(`


  Running an effect without implementation

`, props)
  return new Promise(() => {})
}


