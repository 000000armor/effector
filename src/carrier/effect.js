//@flow

// import {async as subject, type Subject} from 'most-subject'

import {Event} from './event'

import {runEffect} from './run-effect'


export function Effect<
  Params,
  Done,
  Fail,
  State
>(name: string) {
  function runner(typeId, type, payload, dispatch) {
    return runEffect(payload, dispatch, this)
  }
  this.thunk = never
  Event.call(this, name, runner)
  // Object.assign(eff, this)
  // return eff
}

Effect.prototype = Object.create(Event.prototype)
Effect.prototype.constructor = Effect

Effect.prototype.use = function use(thunk: (params: *) => Promise<*>) {
  this.thunk = thunk
  return this
}

function never(props: any): Promise<any> {
  console.warn(`


  Running an effect without implementation

`, props)
  return new Promise(() => {})
}
