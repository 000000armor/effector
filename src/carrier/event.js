//@flow

import {from, of, fromPromise, type Stream} from 'most'
// import type {Subject} from 'most-subject'

// import type {Named, Typed, WithStateLink, Emitter} from '../index.h'
import type {EpicFun} from '../index.h'
import {carrier, type Carrier} from './carrier'
import {type ID, createIDType} from '../id'
import {BareEvent} from './bare-event'
const nextID = createIDType()
// import {registerType} from './register'
// import {on} from './pubsub'
// import {Act, type ActTag} from './act'

import {add} from './register'

export class Event<P, State = any> extends BareEvent<P, Carrier<P>> {
  $call: P => Carrier<P>
  id: ID = nextID()
  used: number = 0
  dispatch: (x: Carrier<P>) => Carrier<P>
  constructor(
    description: string
  ) {
    super()
    const isSerializable = (typeof description === 'string')
      && /^[0-9A-Z_]+$/.test(description)
    if (isSerializable) {
      add(description)
    }
  }

  run(payload: P): Carrier<P> {
    this.used += 1
    return carrier(
      this.id,
      this.getType(),
      payload,
      this.dispatch
    )
  }

  watch<R>(
    handler: (data: P, state: State) => R
  ): Stream<R> {
    const result = this.epic(
      (data$, state$) => watcher(data$, state$, handler)
    )
    // result.observe(e => console.log(e))
    // result.observe(e => this.emit(e))
    return result
  }
  epic: <R>(epic: EpicFun<P, State, R>) => Stream<R>
}

function watcher<P, State, R>(
  data$: Stream<$Exact<{data: P, state: State}>>,
  state$: Stream<State>,
  handler: (data: P, state: State) => R
): Stream<R> {
  return data$
    .map(
      async({data, state}): Promise<Stream<R>> => {
        try {
          const handled = handler(data, state)
          let result: R
          if (handled != null && typeof handled.then === 'function')
            result = await handled
          else
            result = handled
          return of(result)
        } catch (error) {
          console.warn(`\n  Watcher throw\n  ${error && error.message || ''}`)
          const none: R[] = []
          return from(none)
        }
      }
    )
    .map(fromPromise)
    .join()
    .join()
}
