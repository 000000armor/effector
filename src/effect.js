//@flow

import type {Stream} from 'most'
import {async as subject, type Subject} from 'most-subject'

import type {Effect, Event, RawAction} from './index.h'
import {nextPayloadID, nextEventID, toTag, counter, type Tag} from './id'

import {EventConstructor} from './event'
import {safeDispatch, port} from './port'

import {basicCommon} from './event-common'

export function EffectConstructor<State, Params, Done, Fail>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events: Map<string | Tag, Event<any, State>>,
  name: string,
  action$: Subject<Params> = subject()
): Effect<Params, Done, Fail, State> {
  const handlers = new Set

  const {eventID, nextSeq, getType} = basicCommon(domainName, name)

  handlers.add((payload, state) => action$.next(payload))
  // handlers.add((payload, state) => state$.next(state))
  let using: (params: Params) => Promise<Done> = never
  function use(thunk: (params: Params) => Promise<Done>) {
    using = thunk
  }
  const create = (payload: Params) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
    },
  })
  const effect = (params: Params) => {
    let run = () => {
      run = () => {}
      using(params).then(
        result => {
          const data = {params, result}
          dispatch(done(data))
          resolve(data)
        },
        error => {
          const data = {params, error}
          dispatch(fail(data))
          reject(data)
        },
      )
    }
    const runner = (dispatchHook) => {
      if (canDispatch) {
        const dispatcher = dispatchHook
          ? dispatchHook
          : dispatch
        canDispatch = false
        dispatcher(result)
        for (const handler of handlers) {
          handler(params, getState())
        }
        // action$.next(params)
        run()
      }
    }
    let resolve = (val) => {}
    let reject = (val) => {}
    const doneP: Promise<{params: Params, result: Done}> = new Promise(rs => resolve = rs)
    const failP: Promise<{params: Params, error: Fail}> = new Promise(rs => reject = rs)
    const toSend = Promise.resolve(params)
    const result = create(params)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        runner(dispatchHook)
        return toSend
      },
      done() {
        runner()
        return doneP
      },
      fail() {
        runner()
        return failP
      },
      promise() {
        runner()
        return Promise.race([
          doneP,
          failP.then(data => Promise.reject(data)),
        ])
      }
    }
  }
  const done: Event<{params: Params, result: Done}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'done'].join(' ')
  )
  const fail: Event<{params: Params, error: Fail}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'fail'].join(' ')
  )
  function watch<R>(fn: (params: Params, state: State) => R) {
    handlers.add((params, state) => {
      const result = fn(params, state)
      safeDispatch(result, dispatch)
    })
  }

  effect.use = use
  effect.getType = getType
  effect.done = done
  effect.fail = fail
  effect.watch = watch
  effect.epic = port(dispatch, state$, action$)
  return effect
}

function never(props: any): Promise<any> {
  console.warn(`


  Running an effect without implementation

`, props)
  return new Promise(() => {})
}
