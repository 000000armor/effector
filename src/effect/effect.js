//@flow

import {type Stream, combine} from 'most'
import {subject, type Subject} from '../subject'

import type {Effect, Event, RawAction, DomainConfig} from '../index.h'
import {nextPayloadID, nextEventID, toTag, counter, type Tag} from '../id'

import {EventConstructor} from '../event'
import {port} from '../port'

import {basicCommon, observable} from '../event-common'
import {implWarn, implError, watchWarn} from './warn'

export function EffectConstructor<State, Params, Done, Fail>(
  domainName: string,
  dispatch: <T>(value: T) => T,
  getState: () => State,
  state$: Stream<State>,
  events: Map<string | Tag, Event<any, State>>,
  name: string,
  opts: DomainConfig,
  action$: Subject<Params> = subject(),
): Effect<Params, Done, Fail, State> {
  const handlers = new Set()

  const {eventID, nextSeq, getType} = basicCommon(domainName, name)

  handlers.add((payload, state) => action$.next(payload))
  let using: (params: Params) => Promise<Done> = getWarn({
    domainName,
    name,
    mode: () => opts.effectImplementationCheck(),
  })
  function use(thunk: (params: Params) => Promise<Done>) {
    using = thunk
  }
  const create = (payload: Params) => ({
    type: getType(),
    payload,
    meta: {
      index: nextPayloadID(),
      eventID,
      seq: nextSeq(),
      passed: false,
    },
  })
  const effect = (params: Params) => {
    let run = () => {
      run = () => {}
      using(params).then(
        result => {
          const data = {params, result}
          dispatcher(done(data), dispatch)
          resolve(data)
        },
        error => {
          const data = {params, error}
          dispatcher(fail(data), dispatch)
          reject(data)
        },
      )
    }
    const runner = dispatchHook => {
      if (canDispatch) {
        canDispatch = false
        dispatcher(result, dispatch, dispatchHook)
        for (const handler of handlers) {
          handler(params, getState())
        }
        run()
        // action$.next(params)
      }
    }
    let resolve = val => {}
    let reject = val => {}
    const doneP: Promise<{params: Params, result: Done}> = new Promise(
      rs => (resolve = rs),
    )
    const failP: Promise<{params: Params, error: Fail}> = new Promise(
      rs => (reject = rs),
    )
    const toSend = Promise.resolve(params)
    const result = create(params)
    let canDispatch = true
    return {
      ...result,
      raw() {
        return result
      },
      send(dispatchHook) {
        runner(dispatchHook)
        return toSend
      },
      done() {
        runner()
        return doneP
      },
      fail() {
        runner()
        return failP
      },
      promise() {
        runner()
        return Promise.race([doneP, failP.then(data => Promise.reject(data))])
      },
    }
  }
  const done: Event<{params: Params, result: Done}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'done'].join(' '),
  )
  const fail: Event<{params: Params, error: Fail}, State> = EventConstructor(
    domainName,
    dispatch,
    getState,
    state$,
    events,
    [name, 'fail'].join(' '),
  )
  function watch<R>(fn: (params: Params, state: State) => R) {
    effect.epic(
      (data$, state$) => combine(
        (data, state) => ({ data, state }),
        data$, state$
      ).sampleWith(data$)
        .map(({data, state}) => {
          try {
            const result = fn(data, state)
            return result
          } catch (err) {
            watchFailWarn({
              domainName, name,
              mode: opts.watchFailCheck()
            }, err)
          }
        })
    )
  }

  effect.use = use
  effect.getType = getType
  setToString(effect, getType)
  effect.done = done
  effect.fail = fail
  effect.watch = watch
  effect.epic = port(dispatch, state$, action$)
  effect.trigger = function trigger(
    query: (state: State) => Params,
    eventName: string = 'trigger',
  ): Event<void, State> {
    const triggerEvent = EventConstructor(
      domainName,
      dispatch,
      getState,
      state$,
      events,
      [name, eventName].join(' '),
    )
    triggerEvent.watch((_, state) => {
      const queryResult = query(state)
      return effect(queryResult)
    })

    return triggerEvent
  }
  // effect.port = function port<R>(events$: Stream<R>) {
  //   return events$.observe(data => safeDispatch(data, dispatch))
  // }
  effect.subscribe = subscriber => action$.subscribe(subscriber)
  observable(effect, action$)
  return effect
}

function setToString(effect: any, getType) {
  effect.toString = getType
}

function dispatcher(value, dispatchDefault, dispatchHook, isBatch = false) {
  if (!isAction(value)) {
    if (__DEV__) {
      console.warn(`Not an action`, value)
    }
    return value
  }
  const dispatch =
    typeof dispatchHook === 'function' ? dispatchHook : dispatchDefault
  const result = dispatch(value)
  return result
}

function isAction(data): boolean %checks {
  return (
    typeof data === 'object' && data != null && typeof data.type === 'string'
  )
}

function watchFailWarn({domainName, name, mode}, error) {
  switch (mode) {
    case 'throw': {
      throw error
    }
    case 'warn': {
      watchWarn(domainName, name, error)
    }
    case 'off':
    default: {
      return
    }
  }
}

const getWarn = ({domainName, name, mode}) =>
  function never(props: any): Promise<any> {
    switch (mode()) {
      case 'throw': {
        return Promise.reject(implError(domainName, name, props))
      }
      case 'warn': {
        implWarn(domainName, name, props)
      }
      case 'off':
      default: {
        return new Promise(() => {})
      }
    }
  }
