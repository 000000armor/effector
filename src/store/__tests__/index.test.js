//@flow

import {createStore, createApi, restore, createStoreObject} from '..'
import {createEvent} from 'effector/event'
import {createEffect} from 'effector/effect'
import {spy, getSpyCalls} from 'effector/fixtures'

test('.map', () => {
  const newWord = createEvent<string>()
  const a = createStore('word').on(newWord, (_, word) => word)

  const b = a.map(word => word.length)

  const sum = b.map((ln, prevLn) => ln + prevLn, 0)

  sum.watch(spy)

  expect(a.getState()).toBe('word')
  expect(b.getState()).toBe(4)
  expect(sum.getState()).toBe(4)

  newWord('lol')

  expect(a.getState()).toBe('lol')
  expect(b.getState()).toBe(3)
  expect(sum.getState()).toBe(7)

  newWord('long word')

  expect(a.getState()).toBe('long word')
  expect(b.getState()).toBe(9)
  expect(sum.getState()).toBe(16)

  expect(spy).toHaveBeenCalledTimes(3)

  newWord('')

  expect(spy).toHaveBeenCalledTimes(3)
})

describe('.watch', () => {
  it('supports functions', () => {
    const newWord = createEvent<string>()
    const a = createStore('word').on(newWord, (_, word) => word)

    const b = a.map(word => word.length)

    const sum = b.map((ln, prevLn) => ln + prevLn, 0)

    sum.watch(spy)

    newWord('lol')

    newWord('long word')

    expect(spy).toHaveBeenCalledTimes(3)

    newWord('')

    expect(spy).toHaveBeenCalledTimes(3)
  })
  it('returns unsubscribe function', () => {
    const newWord = createEvent<string>()
    const a = createStore('word').on(newWord, (_, word) => word)

    const b = a.map(word => word.length)

    const sum = b.map((ln, prevLn) => ln + prevLn, 0)

    const unsub = sum.watch(sum => {
      console.warn('summ called', sum)
      spy(sum)
    })

    newWord('lol')

    newWord('long word [1]')

    expect(spy).toHaveBeenCalledTimes(3)

    unsub()

    newWord('long word _ [2]')
    expect(spy).toHaveBeenCalledTimes(3)
  })
  it('supports events', () => {
    const newWord = createEvent<string>('new word')
    const spyEvent = createEvent('spy event')
    const a = createStore('word').on(newWord, (_, word) => word)

    const b = a.map(word => word.length)

    const sum = b.map((ln, prevLn) => ln + prevLn, 0)

    sum.watch(spyEvent, (store, event) => spy(store, event))

    newWord('lol')
    expect(spy).toHaveBeenCalledTimes(0)
    spyEvent(1)
    spyEvent(2)
    expect(spy).toHaveBeenCalledTimes(2)

    newWord('')
    expect(spy).toHaveBeenCalledTimes(2)
    newWord(' ')
    expect(spy).toHaveBeenCalledTimes(2)

    spyEvent(3)
    newWord('long word')
    expect(spy).toHaveBeenCalledTimes(3)
    expect(getSpyCalls()).toEqual([[7, 1], [7, 2], [8, 3]])
  })
  it('supports effects', () => {
    const newWord = createEvent<string>('new word')
    const spyEvent = createEffect('spy effect')
    spyEvent.use(args => (console.log(args), args))
    const a = createStore('word').on(newWord, (_, word) => word)

    const b = a.map(word => word.length)

    const sum = b.map((ln, prevLn) => ln + prevLn, 0)

    sum.watch(spyEvent, (store, event) => spy(store, event))

    newWord('lol')
    expect(spy).toHaveBeenCalledTimes(0)
    spyEvent(1)
    spyEvent(2)
    expect(spy).toHaveBeenCalledTimes(2)

    newWord('')
    expect(spy).toHaveBeenCalledTimes(2)
    newWord(' ')
    expect(spy).toHaveBeenCalledTimes(2)

    spyEvent(3)
    newWord('long word')
    expect(spy).toHaveBeenCalledTimes(3)
    expect(getSpyCalls()).toEqual([[7, 1], [7, 2], [8, 3]])
  })
})

test('.off', () => {
  const newWord = createEvent<string>()
  const a = createStore('word').on(newWord, (_, word) => word)

  const b = a.map(word => word.length)

  const sum = b.map((ln, prevLn) => ln + prevLn, 0)

  sum.watch(spy)

  expect(a.getState()).toBe('word')
  expect(b.getState()).toBe(4)
  expect(sum.getState()).toBe(4)

  newWord('lol')

  expect(a.getState()).toBe('lol')
  expect(b.getState()).toBe(3)
  expect(sum.getState()).toBe(7)

  a.off(newWord)

  newWord('long word')

  expect(a.getState()).toBe('lol')
  expect(b.getState()).toBe(3)
  expect(sum.getState()).toBe(7)

  expect(spy).toHaveBeenCalledTimes(2)

  newWord('')

  expect(spy).toHaveBeenCalledTimes(2)
})

test('create api', () => {
  const shape = restore({
    todos: ['to do', 'list'],
    newTodo: '',
    selected: [1],
  })
  const full = createStoreObject(shape)
  const listCmd = createApi(shape.todos, {
    add(todos, item) {
      return [...todos, item]
    },
    remove(todos, index) {
      return [...todos].splice(index, 1)
    },
  })
  shape.selected.on(listCmd.remove, (selected, index) => {
    const pos = selected.indexOf(index)
    if (pos === -1) return selected
    return [...selected].splice(pos, 1)
  })
  const selectionCmd = createApi(shape.selected, {
    add(selected, item) {
      return [...selected, item]
    },
    remove(selected, index) {
      return [...selected].splice(index, 1)
    },
  })
  expect(full.getState()).toMatchObject({
    todos: ['to do', 'list'],
    newTodo: '',
    selected: [1],
  })
  listCmd.add('new todo')
  selectionCmd.add(2)
  expect(full.getState()).toMatchObject({
    todos: ['to do', 'list', 'new todo'],
    newTodo: '',
    selected: [1, 2],
  })
})
