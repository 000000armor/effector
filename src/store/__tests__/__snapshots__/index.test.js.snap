// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createStore 1`] = `"createStore: First argument can't be undefined, use null instead."`;
